hoomd/mpcd/ConcentricCylindersGeometry.h

```diff
@@ -0,0 +1,170 @@
+
+class __attribute__((visibility("default"))) ConcentricCylindersGeometry
+    {
+    public:
+    //! Constructor
+    /*!
+     * \param R0 Inner radius of the cylinder
+     * \param R1 Outer radius of the cylinder
+     * \param angular_speed Angular speed of the outer cylinder
+     * \param no_slip Boundary condition at the wall (slip or no-slip)
+     */
+    HOSTDEVICE ConcentricCylindersGeometry(Scalar R0, Scalar R1, Scalar angular_speed, bool no_slip)
+        : m_R0_sq(R0 * R0), m_R1_sq(R1 * R1), m_w(angular_speed), m_no_slip(no_slip)
+        {
+        }
+
+    //! Detect collision between the particle and the boundary
+    /*!
+     * \param pos Proposed particle position
+     * \param vel Proposed particle velocity
+     * \param dt Integration time remaining
+     *
+     * \returns True if a collision occurred, and false otherwise
+     *
+     * \post The particle position \a pos is moved to the point of reflection, the velocity \a vel
+     * is updated according to the appropriate bounce back rule, and the integration time \a dt is
+     * decreased to the amount of time remaining.
+     */
+    HOSTDEVICE bool detectCollision(Scalar3& pos, Scalar3& vel, Scalar& dt) const
+        {
+        const Scalar rsq = pos.x * pos.x + pos.y * pos.y;
+        /*
+         * Check if particle is in bounds
+         */
+        const signed char sign = (char)((rsq < m_R0_sq) - (rsq > m_R1_sq));
+        // exit immediately if no collision is found or particle is not moving normal to the wall
+        // (since no new collision could have occurred if there is no normal motion)
+        if (sign == 0 || vel.x * vel.x + vel.y * vel.y == Scalar(0))
+            {
+            dt = Scalar(0);
+            return false;
+            }
+
+        /*
+         * Find the time remaining when the particle is collided with wall. This time is computed
+         * by backtracking the position, dot(pos, pos) = R^2.
+         */
+        const Scalar Rsq = (sign == 1) ? m_R0_sq : m_R1_sq;
+        const Scalar a = vel.x * vel.x + vel.y * vel.y;
+        const Scalar b = -Scalar(2) * (pos.x * vel.x + pos.y * vel.y);
+        const Scalar c = pos.x * pos.x + pos.y * pos.y - Rsq;
+        dt = (-b + sign * (slow::sqrt(b * b - Scalar(4) * a * c))) / (Scalar(2) * a);
+
+        // backtrack the particle for dt to get to point of contact
+        pos -= vel * dt;
+
+        // update velocity according to boundary conditions
+        // no-slip requires reflection of the tangential components.
+        if (m_no_slip)
+            {
+            vel = -vel;
+            if (sign == -1)
+                {
+                vel.x += Scalar(2) * m_w * -pos.y;
+                vel.y += Scalar(2) * m_w * pos.x;
+                }
+            }
+        // slip requires reflection of the normal components
+        else
+            {
+            vel.x -= Scalar(2) * pos.x * pos.x * vel.x / (Rsq);
+            vel.y -= Scalar(2) * pos.y * pos.y * vel.y / (Rsq);
```

Comment:
A short comment here on where these formulas come from would also be helpful for future!
```suggestion
            vel.x -= Scalar(2) * pos.x * pos.x * vel.x / Rsq;
            vel.y -= Scalar(2) * pos.y * pos.y * vel.y / Rsq;
```